aten/src/ATen/CPUApplyUtils.h

```diff
@@ -0,0 +1,271 @@
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor2, dim, 1)
+  auto t1_numel = tensor1.numel();
+  auto t2_numel = tensor2.numel();
+  if(t1_numel != t2_numel) {
+    std::ostringstream oss;
+    oss << "inconsistent tensor size, expected " << tensor1.sizes() << " and " << tensor2.sizes()
+        << " to have the same number of elements, but got " << t1_numel << " and " << t2_numel << " elements respectively";
+    throw std::runtime_error(oss.str());
+  }
+  while(!TH_TENSOR_APPLY_hasFinished)
+  {
+    /* Loop through the inner most region of the Tensor */
+    for(; tensor1_i < tensor1_size && tensor2_i < tensor2_size; tensor1_i++, tensor2_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride)
+    {
+      op(*tensor1_data, *tensor2_data, TH_TENSOR_APPLY_hasFinished);
+    }
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
+  }
+  if(tensor1_counter != NULL)
+    delete [] tensor1_counter;
+  if(tensor2_counter != NULL)
+    delete [] tensor2_counter;
+}
+
+template<typename ScalarType, typename Op>
+void CPU_tensor_apply2(Tensor tensor1, Tensor tensor2, Op op) {
+  CPU_tensor_apply2_dim<ScalarType, Op>(tensor1, tensor2, -1, op);
+}
+
+template <typename ScalarType, typename Op>
+void CPU_tensor_apply3_dim(Tensor &tensor1, Tensor& tensor2, Tensor& tensor3, int64_t dim, Op op) {
+  check_correct_backend(tensor1, tensor2, tensor3);
+  bool TH_TENSOR_APPLY_hasFinished = false;
+  int64_t TH_TENSOR_dim_index = 0;
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor1, dim, 1)
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor2, dim, 1)
+  __ATH_TENSOR_APPLYX_PREAMBLE(ScalarType, tensor3, dim, 1)
+
+  int elements_equal = 1;
+  auto t1_numel = tensor1.numel();
+  auto t2_numel = tensor2.numel();
+  auto t3_numel = tensor3.numel();
+  if(t1_numel!= t2_numel) {
+    elements_equal = 0;
+  } else if(t1_numel != t3_numel) {
+    elements_equal = 0;
+  }
+  if (elements_equal == 0) {
+    std::ostringstream oss;
+    oss << "inconsistent tensor size, expected " << tensor1.sizes() << ", " << tensor2.sizes() << ", and " << tensor3.sizes()
+        << " to have the same number of elements, but got " << t1_numel << ", " << t2_numel << ", and " << t3_numel << " elements respectively";
+    throw std::runtime_error(oss.str());
+  }
+
+  while(!TH_TENSOR_APPLY_hasFinished)
+  {
+    /* Loop through the inner most region of the Tensor */
+    for(; tensor1_i <  tensor1_size && tensor2_i < tensor2_size && tensor3_i < tensor3_size; tensor1_i++, tensor2_i++, tensor3_i++, tensor1_data += tensor1_stride, tensor2_data += tensor2_stride, tensor3_data += tensor3_stride)
+    {
+      op(*tensor1_data, *tensor2_data, *tensor3_data, TH_TENSOR_APPLY_hasFinished);
+    }
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor1, 0)
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor2, 0)
+    __ATH_TENSOR_APPLYX_UPDATE_COUNTERS(tensor3, 0)
+  }
+  if(tensor1_counter != NULL)
+    delete [] tensor1_counter;
```

Comment:
It would be nice to use modern C++ here, but I'd rather have this merged and available and then improve the implementation later.

I think this may leak the counters on exceptions. We could avoid that with `std::unique_ptr` or even `std::vector`.

Reply:
Yes, agreed.